docs/_includes/Statements.md

```diff
@@ -239,6 +239,153 @@ Note that the form
 
 is diagnosed as a label in which the user forgot the **label** keyword.
 
+## Update with Failure Statement
+
+An update statement using `:-`   instead of `:=` exits execution of a method immediately if a failure is reported. This is a form of exceptional return from the calling method.
+
+To use this form of update,
+   * the caller must have a first output value whose type is failure-compatible
+   * if the RHS of the update-with-failure statement is a method call, the first output value of the callee must be failure-compatible
+   * if instead the RHS of the update-with-failure statement is one or more expressions, the first of these expressions must be a value with a failure-compatible type
+
+A failure-compatible type is a type that has the following:
+   * a predicate `predicate IsFailure()`
+   * a method `method PropagateFailure() returns _X_ `: The type _X_ must be a failure-compatible type and must be the type of the first output of the caller.
+   * a method `method Extract() returns _T_` where _T_ is some type for the value carried by the type.
+
+TODO: Why should PropagateFailure and Extract be functions?
```

Comment:
In the proposed translation of `:-` in https://github.com/dafny-lang/dafny/issues/603, each of `PropagateFailure` and `Extract` can be either a function or a method. In either case, it should have no in-parameters (other than the receiver parameter). If it is a method, then it should have exactly one out-parameter.

There are two forms of failure-compatible types. One has an `Extract` member, the other does not. If the type has an `Extract` member, then it must be a function or method.

The type `_X_` can be anything. It need not be a failure-compatible type. (To be used with `:-` in a particular context, the type `_X_` must be assignable to the first out-parameter of the enclosing method, but that need not be a failure-compatible type.)

Rather than talking about `_X_` and `_T_` above (which isn't syntactically correct, anyway), I suggest simply saying:

* a compiled (that is, non-ghost) predicate (that is, function with a `bool` result type) `IsFailure()`
* a compiled function or compiled 1-out-parameter method `PropagateFailure()`
* either no member called `Extract`, or a function or compiled 1-out-parameter method `Extract()`

Note that `Extract()` can be either compiled or ghost, but it must agree with the first LHS of the `:-` statement.

Reply:
Revised